sympy/matrices/eigen.py

```diff
@@ -372,6 +373,76 @@ def _is_diagonalizable(M, reals_only=False, **kwargs):
     return _is_diagonalizable_with_eigen(M, reals_only=reals_only)[0]
 
 
+def _householder_vector(x, precision):
+    if not x.cols == 1:
+        raise ValueError("Input must be a column matrix")
+    alp = x[1:, 0].T * x[1:, 0]
+    sig = alp[0, 0]
+    v = x.copy()
+    v[0, 0] = 1
+    if precision is not None:
+        sig = N(sig)
+    if N(sig) == 0 and x[0] >= 0:
+        bet = 0
+    elif N(sig) == 0 and x[0] < 0:
+        bet = -2
+    else:
+        mu = sqrt(x[0]*x[0] + sig)
+        if x[0] <= 0:
+            v[0] = x[0] - mu
+        else:
+            v[0] = -sig/(x[0]+mu)
+        bet = 2 * v[0] * v[0] /( sig + v[0]*v[0] )
+        v = v/v[0]
+    return v, bet
+
+
+def _bidiag_hholder(M, precision):
+    m = M.rows
+    n = M.cols
+    A = M.copy()
+    for i in range(min(m, n)):
+        v, bet = _householder_vector(A[i:, i], precision)
+        A[i:, i:] = (M.eye(m-i) - bet * v * v.T)*A[i:, i:]
+        if i + 1 <= n - 2:
+            v, bet = _householder_vector(A[i, i+1:].T, precision)
+            A[i:, i+1:] = A[i:, i+1:]*(M.eye(n-i-1)- bet*v*v.T)
+    if precision is not None:
+        for row in range(A.rows):
+            for col in range(A.cols):
+                A[row, col] = round(A[row, col], precision)
+
+    return A
+
+
+def _bidiagonalize(M, precision=None, upper=True):
+    """
+    Returns bidiagonalized form of the input Matrix.
+
+    Parameters
+    ==========
+
+    precision : int. Number of digits beside decimal point.
+                Default: None, which makes all the computation to be done symbollically
```

Comment:
I don't think that it's actually transparent inside how `precision` is supposed to work.
Like how this guarantees the precision
```
    if precision is not None:
        sig = N(sig)
```

And even I have some questions whether this keyword is needed or not. For example, you can first convert the matrix to floating point and then all the operations can be done with floating point arithmetics.

Reply:
How should I convert the Matrix to  floating point? Like, is there any recommended or suggested way to convert a Sympy Matrix to floating-point.

And even if I convert the entire Matrix to floating-point, some error is always introduced of the order of 10^-16, what should I do regarding that?

Reply:
`Matrix.evalf` can have the logic. 
I think that a zero test using epsilon can only make sense for purely floating point matrix, so you may better handle that automatically after `all(isinstance(x, float) for x in M)`.